pkg/antctl/raw/multicluster/join.go

```diff
@@ -0,0 +1,284 @@
+		joinOpts.MemberClusterOpt.ClusterSetID)
+	secret := &v1.Secret{}
+	if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: joinOpts.LeaderClusterOpt.Secret, Namespace: joinOpts.MemberClusterOpt.Namespace}, secret); err != nil {
+		fmt.Fprintf(cmd.OutOrStdout(), "Failed to get secret \"%s\" in Namespace %s: %v\n", joinOpts.LeaderClusterOpt.Secret, joinOpts.MemberClusterOpt.Namespace, err)
+		return err
+	}
+
+	crtData, token, err := commonarea.GetSecretCACrtAndToken(secret)
+	if err != nil {
+		return err
+	}
+	config, err := clientcmd.BuildConfigFromFlags(joinOpts.LeaderClusterOpt.APIServer, "")
+	if err != nil {
+		return err
+	}
+	config.BearerToken = string(token)
+	config.CAData = crtData
+
+	remoteClient, err := client.New(config, client.Options{Scheme: multiclusterscheme.Scheme})
+	if err != nil {
+		return err
+	}
+
+	memberClusterAnnounce := "member-announce-from-" + joinOpts.MemberClusterOpt.ClusterID
+	if err := waitForMemberClusterAnnounceReady(remoteClient, memberClusterAnnounce, joinOpts.LeaderClusterOpt.Namespace); err != nil {
+		fmt.Fprintf(cmd.OutOrStdout(), "Failed to wait for MemberClusterAnnounce \"%s\" in Namespace %s in leader cluster: %v\n", memberClusterAnnounce, joinOpts.LeaderClusterOpt.Namespace, err)
+		return err
+	}
+
+	if err := waitForLeaderClusterSetReady(k8sClient, joinOpts.MemberClusterOpt.ClusterSetID, joinOpts.MemberClusterOpt.Namespace, joinOpts.LeaderClusterOpt.ClusterID); err != nil {
+		fmt.Fprintf(cmd.OutOrStdout(), "Failed to wait for ClusterSet \"%s\" in Namespace %s in member cluster: %v\n", joinOpts.MemberClusterOpt.ClusterSetID, joinOpts.MemberClusterOpt.Namespace, err)
+		return err
+	}
+
+	return nil
+}
+
+func waitForMemberClusterAnnounceReady(client client.Client, memberClusterAnnounce string, namespace string) error {
+	return wait.PollImmediate(
+		1*time.Second,
+		30*time.Second,
+		func() (bool, error) {
+			announce := &multiclusterv1alpha1.MemberClusterAnnounce{}
+			if err := client.Get(context.TODO(), types.NamespacedName{Name: memberClusterAnnounce, Namespace: namespace}, announce); err != nil {
+				if apierrors.IsNotFound(err) {
+					return false, nil
+				}
+				return false, err
+			}
+			return announce.DeletionTimestamp == nil, nil
+		})
+}
+
+func waitForLeaderClusterSetReady(client client.Client, name string, namespace string, clusterID string) error {
+	return wait.PollImmediate(
+		1*time.Second,
+		3*time.Minute,
+		func() (bool, error) {
+			clusterSet := &multiclusterv1alpha1.ClusterSet{}
+			if err := client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, clusterSet); err != nil {
+				if apierrors.IsNotFound(err) {
+					return false, nil
+				}
+				return false, err
+			}
+
+			for _, status := range clusterSet.Status.ClusterStatuses {
+				if status.ClusterID == clusterID {
+					for _, cond := range status.Conditions {
+						if cond.Type == multiclusterv1alpha1.ClusterReady {
```

Comment:
We do update the local cluster status in ClusterSet in the member cluster (note the command works on the member not the leader cluster), right? Please verify the command works well.

Reply:
The functions will check the ClusterSet in the member cluster so it works well.